### PR TITLE
Load LOPs: Support update via context options (replace parm callbacks with expressions)

### PR DESCRIPTION
**Pull request by BigRoy**
## Changelog Description

Replace the Load LOPs parm callbacks with expressions on the file parms to support Solaris Context Options.

## Additional info

With this PR, the LOP load nodes now LIVE update based on context options. Meaning, using context-options you can now easily do multi-shot lighting!

Fixes: https://github.com/ynput/ayon-houdini/issues/111

https://github.com/user-attachments/assets/a3807f41-eb24-4da9-94b9-1a88fca840db

_Credit: Audio snippet in the background at the beginning is a few seconds of the start from this [Youtube video](https://www.youtube.com/watch?v=DeXoACwOT1o)_

#### TODO

- [x] Fix the 'warning message' remaining on the node on path update because it doesn't trigger that callback anymore
- [x] Fix the thumbnails (when enabled) not updating on representation changes.
- [x] Add "refresh thumbnail" button.

## Testing notes:

1. [x] Load Asset LOP should work consistently
2. [x] Load Shot LOP should work consistently
3. [x] Both should NOT be super slow
4. [x] When using Solaris Context Options the loader should update along live. 
    - [x] even with Edit Context Options LOP updates should be live

### CODE REVIEW

**Review by MustafaJafar — APPROVED** [review]
This is cool and it's live!
Animation_7

Should we care about these warnings ? 
image

Also, should we have default context options and probably manage them on AYON context change?

**Comment by BigRoy** [comment]
> Should we care about these warnings ?

Yes, we should! I thought I had fixed that already but maybe I forgot to push. Will continue with this.

Will also need to decide what to do with the thumbnails. Before this PR it was created to "auto-update" on context update (but especially now with those being expressions that makes even less sence.) Mark from SideFX says it best: 
> [You should really not be "doing things" as a result of evaluating expressions.](https://www.sidefx.com/forum/topic/98147/?page=1#post-431580)

So I guess for now the easiest way is to remove how it's "live" and then just add a "Update thumbnail" button. We can think of better fixes down the line.

**Comment by MustafaJafar** [comment]
Isn't this cool!
I published three products `pointcacherock_1`, `pointcacherock_2` and `pointcacherock_3`
and loading them inside a loop works! (as loops are based on context options)

_Note: I don't know if my example here is favorite in production. tbh, I don't know what is the best way to publish multiple variants of my pointcache products. maybe it's the correct workflow since different variants of an AYON product are just different AYON products._

image

**Comment by BigRoy** [comment]
@MustafaJafar could you give it one last test-drive?

**Comment by MustafaJafar** [comment]
on my side, 
- when I click `Reload Files` then `Update Thumbnail`, I get the following warning.
image
- when I click `Clear Cache` then `Update Thumbnail`, I get the following error: 
```python
Traceback (most recent call last):
  File "ayon::Lop/lop_import::1.0/update_thumbnail", line 1, in <module>
  File "E:\Ynput\ayon-houdini\client\ayon_houdini\api\hda_utils.py", line 189, in update_thumbnail
    repre_entity = get_representation_by_id(project_name, representation_id)
  File "E:\Ynput\ayon-launcher\.venv\lib\site-packages\ayon_api\_api.py", line 3449, in get_representation_by_id
    return con.get_representation_by_id(*args, **kwargs)
  File "E:\Ynput\ayon-launcher\.venv\lib\site-packages\ayon_api\server_api.py", line 6533, in get_representation_by_id
    for representation in representations:
  File "E:\Ynput\ayon-launcher\.venv\lib\site-packages\ayon_api\server_api.py", line 6489, in get_representations
    for parsed_data in query.continuous_query(self):
  File "E:\Ynput\ayon-launcher\.venv\lib\site-packages\ayon_api\graphql.py", line 378, in continuous_query
    raise GraphQlQueryFailed(
ayon_api.exceptions.GraphQlQueryFailed: GraphQl query Failed: Invalid entity ID None on item 'project/representations' (Line 3 Column 5)
```

File: client/ayon_houdini/api/hda_utils.py [hunk 1]
```diff
@@ -155,47 +172,21 @@ def _get_thumbnail(project_name: str, version_id: str, thumbnail_dir: str):
...
 
-def set_representation(node, representation_id: str):
-    file_parm = node.parm("file")
+def update_thumbnail(node):
+    if not node.evalParm("show_thumbnail"):
+        lib.remove_all_thumbnails(node)
+        return
+
+    representation_id = node.evalParm("representation")
     if not representation_id:
```
Comment (MustafaJafar) [inline client/ayon_houdini/api/hda_utils.py:181]: After digging, I don't know, why the `representation_id` is equal to the lateral `"None"` after clearing the cache. 
Anyways, this works.
```suggestion
    if not representation_id or representation_id == "None":
```

Comment (BigRoy) [reply client/ayon_houdini/api/hda_utils.py:181]: Thanks detective! This seems to hint to another issue that I'd rather fix so I'll try and reproduce and fix it.

Comment (BigRoy) [reply client/ayon_houdini/api/hda_utils.py:181]: Fixed with https://github.com/ynput/ayon-houdini/pull/112/commits/8d85e2638c48089e8fb11e9166f043dbcc5710ce

It took me way longer than I want to admit to find that :') haha..

I did learn that I can quickly update the `hda` code continuously so I don't need to re-run houdini each time. I put e.g. this in a Houdini shelf:
```python
import hou
from ayon_houdini.api import lib, hda_utils
import importlib

# Enforce reload of AYON houdini libs
importlib.reload(lib)
importlib.reload(hda_utils)

# Reload HDAs
hou.hda.reloadAllFiles(rescan=True)

# Ensure the HDA module uses the reloaded libs
for node in hou.selectedNodes():
    hou.hda.reloadHDAModule(node.hdaModule())
```

**Review by MustafaJafar — COMMENTED** [review]
Here you are some test runs: 
I like how it can be used in a for loop.
Animation_23
I ca also use it in a custom HDA.
Animation_25
Update thumbnail is buggy, to make it reliable, I need to trigger reload every time before clicking update thumbnail.
Animation_24

**Comment by BigRoy** [comment]
> Update thumbnail is buggy, to make it reliable, I need to trigger reload every time before clicking update thumbnail.

Correct, update thumbnail is intentionally not live - since expressions should not be allowed to edit the graph aside of the return value from the expression.

As such, @MustafaJafar I suppose everything works? :)

**Comment by MustafaJafar** [comment]
> Correct, update thumbnail is intentionally not live - since expressions should not be allowed to edit the graph aside of the return value from the expression.

yes, but I've to click to button to actually update the thumbnail.

**Pull request merged**